pkg/reconciler/awscodecommitsource/adapter.go

```diff
@@ -43,18 +46,35 @@ type adapterConfig struct {
 	configs source.ConfigAccessor
 }
 
-// adapterDeploymentBuilder returns an AdapterDeploymentBuilderFunc for the
-// given source object and adapter config.
-func adapterDeploymentBuilder(src *v1alpha1.AWSCodeCommitSource, cfg *adapterConfig) common.AdapterDeploymentBuilderFunc {
-	return func(sinkURI *apis.URL) *appsv1.Deployment {
-		return common.NewAdapterDeployment(src, sinkURI,
-			resource.Image(cfg.Image),
-
-			resource.EnvVar(common.EnvARN, src.Spec.ARN.String()),
-			resource.EnvVar(envBranch, src.Spec.Branch),
-			resource.EnvVar(envEventTypes, strings.Join(src.Spec.EventTypes, ",")),
-			resource.EnvVars(common.MakeSecurityCredentialsEnvVars(src.Spec.Credentials)...),
-			resource.EnvVars(cfg.configs.ToEnvVars()...),
-		)
+// Verify that Reconciler implements common.AdapterDeploymentBuilder.
+var _ common.AdapterDeploymentBuilder = (*Reconciler)(nil)
+
+// BuildAdapter implements common.AdapterDeploymentBuilder.
+func (r *Reconciler) BuildAdapter(src v1alpha1.EventSource, sinkURI *apis.URL) *appsv1.Deployment {
+	typedSrc := src.(*v1alpha1.AWSCodeCommitSource)
+
+	return common.NewAdapterDeployment(src, sinkURI,
+		resource.Image(r.adapterCfg.Image),
+
+		resource.EnvVar(common.EnvARN, typedSrc.Spec.ARN.String()),
+		resource.EnvVar(envBranch, typedSrc.Spec.Branch),
+		resource.EnvVar(envEventTypes, strings.Join(typedSrc.Spec.EventTypes, ",")),
+		resource.EnvVars(common.MakeSecurityCredentialsEnvVars(typedSrc.Spec.Credentials)...),
+		resource.EnvVars(r.adapterCfg.configs.ToEnvVars()...),
+	)
+}
+
+// RBACOwners implements common.AdapterDeploymentBuilder.
```

Comment:
I was thinking about it as well but the lack of generics made it challenging.
I can try to give it another shot, maybe it's possible with some clever use of interfaces.

Reply:
For some reason, I was thinking of a common `RBACOwners` function in  `pkg/reconciler/common` that would be called by each source's handler as a workaround.

Reply:
Yes I would love that, but each source has its own Lister type (SNS has a Lister of type SNSLister that contains SNS sources, etc
.).
That's where I'm stuck. Due to the absence of generics, it's difficult to have one function for all.
I'll try to find a way, it won't be elegant but it might work.

Reply:
I was thinking of something along the lines of passing in the Reconciler, but realized `Reconciler` is defined per adapter. Shame there isn't a way to duck type an interface that requires a lister defined as `srcLister` that can be passed in.

Reply:
Yeah. What I can do is create a helper that accepts a callback function with this signature as argument:

```go
func() ([]kmeta.OwnerRefable, error)
```

And inside the helper, just call the function, which wraps whatever Lister was put inside by the reconciler.

That is actually not as hacky as I imagined. On it.

Reply:
Scratch that. It won't work, because listers return strongly typed objects which have to be turned into interfaces, and that can't be done without repeating code :(

Reply:
Guess it can't be helped in any case.